client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -210,53 +213,36 @@ void genServerToBlocks(
       int replicaNum,
-      List<ShuffleServerInfo> excludeServers,
+      Collection<ShuffleServerInfo> excludeServers,
       Map<ShuffleServerInfo, Map<Integer, Map<Integer, List<ShuffleBlockInfo>>>> serverToBlocks,
       Map<ShuffleServerInfo, List<Long>> serverToBlockIds,
       boolean excludeDefectiveServers) {
     if (replicaNum <= 0) {
       return;
     }
-    int partitionId = sbi.getPartitionId();
-    int shuffleId = sbi.getShuffleId();
-    int assignedNum = 0;
-    for (ShuffleServerInfo ssi : serverList) {
-      if (excludeDefectiveServers && replica > 1 && defectiveServers.contains(ssi)) {
-        continue;
-      }
-      if (CollectionUtils.isNotEmpty(excludeServers) && excludeServers.contains(ssi)) {
-        continue;
-      }
-      if (!serverToBlockIds.containsKey(ssi)) {
-        serverToBlockIds.put(ssi, Lists.newArrayList());
-      }
-      serverToBlockIds.get(ssi).add(sbi.getBlockId());
-
-      if (!serverToBlocks.containsKey(ssi)) {
-        serverToBlocks.put(ssi, Maps.newHashMap());
-      }
-      Map<Integer, Map<Integer, List<ShuffleBlockInfo>>> shuffleIdToBlocks = serverToBlocks.get(ssi);
-      if (!shuffleIdToBlocks.containsKey(shuffleId)) {
-        shuffleIdToBlocks.put(shuffleId, Maps.newHashMap());
-      }
 
-      Map<Integer, List<ShuffleBlockInfo>> partitionToBlocks = shuffleIdToBlocks.get(shuffleId);
-      if (!partitionToBlocks.containsKey(partitionId)) {
-        partitionToBlocks.put(partitionId, Lists.newArrayList());
-      }
-      partitionToBlocks.get(partitionId).add(sbi);
-      if (excludeServers != null) {
-        excludeServers.add(ssi);
-      }
-      assignedNum++;
-      if (assignedNum >= replicaNum) {
-        break;
-      }
+    Stream<ShuffleServerInfo> servers;
+    if (excludeDefectiveServers && CollectionUtils.isNotEmpty(defectiveServers)) {
+      servers = Stream.concat(serverList.stream().filter(x -> !defectiveServers.contains(x)),
+          serverList.stream().filter(defectiveServers::contains));
+    } else {
+      servers = serverList.stream();
+    }
+    if (excludeServers != null) {
+      servers = servers.filter(x -> !excludeServers.contains(x));
     }
 
-    if (assignedNum < replicaNum && excludeDefectiveServers) {
-      genServerToBlocks(sbi, serverList, replicaNum - assignedNum,
-          excludeServers, serverToBlocks, serverToBlockIds, false);
+    List<ShuffleServerInfo> selected = servers.limit(replicaNum).collect(Collectors.toList());
+    for (ShuffleServerInfo ssi : selected) {
+      serverToBlockIds.computeIfAbsent(ssi, id -> Lists.newArrayList())
+          .add(sbi.getBlockId());
+      serverToBlocks.computeIfAbsent(ssi, id -> Maps.newHashMap())
+          .computeIfAbsent(sbi.getShuffleId(), id -> Maps.newHashMap())
+          .computeIfAbsent(sbi.getPartitionId(), id -> Lists.newArrayList())
+          .add(sbi);
+    }
+    if (excludeServers != null) {
+      excludeServers.addAll(selected);
```

Comment:
How about put this logic after `limit` of the follow line?
https://github.com/apache/incubator-uniffle/blob/ffc93e6bf8fde26ce668ad20e20a25d826fc01a5/client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java#L235

Reply:
I mean `servers.limit(replicaNum).forEach((server) -> {excludeServers.add(server)}).collect(Collectors.toList());`

Reply:
OK, let's just avoid collect to list.